### DISCUSSION TITLE
Project: 🌇 Adopt generic class based views across the Wagtail admin

### DISCUSSION THREAD

**Discussion by thibaudcolas** [wagtail/outreachy#4] [Ideas]
Please use this Discussions thread to ask anything and everything relating to our [Adopt generic class based views across the Wagtail admin](https://github.com/wagtail/outreachy/blob/main/project-ideas.md#adopt-generic-class-based-views-across-the-wagtail-admin) project idea.

Don’t hesitate to answer other people’s questions and mark this as a contribution on the Outreachy website if you’re particularly happy with your answer.

**Comment by Chiemezuo** [comment]
Hi @thibaudcolas, I've been reading through a lot of the code from PR's (past and present) related to this umbrella project. I just wanted to ask (and do pardon me if the answer seems obvious): aside from the code reuse and customization that switching to class-based views will bring, are there performance benefits over function-based views?
Can we say there will be performance boosts after the switch? (Or even drawbacks) Or is it just the development benefits?

**Reply by lb-** [reply]
The performance differences should be negligible, but it's probably worth researching CBV vs FBV performance to see if there are any gotchas to consider. 

The main benefits I see are:

1. Customisations - Wagtail allows you to override URLs easily and templates easily but it's near impossible to change how the internals of an FBV work based on importing the 'exisiting' one. CBVs make it much easier to just customise one part. 
2. Consistency - As we do more with Snippets and the ModelViewSets, it's much easier to support code that behaves the same throughout Wagtail. It's also much easier to document the supported use cases if it all works a similar way. Irrespective of what module.

**Reply by Chiemezuo** [reply]
Wow! This really clarified things. I'll do some extra research on it. Thank you @lb-

**Reply by thibaudcolas** [reply]
I wouldn’t expect any performance benefits, unless this allows us to achieve particular refactorings that improve performance (for example permissions).

**Comment by meli-imelda** [comment]


**Reply by Chiemezuo** [reply]
Yes. The people here are really supportive and offer guidance regardless.

**Comment by Agaba-derrick** [comment]
I am also still trying to quickly understand the internals of Wagtail as per the untold requirement of this project. Still, I also think that even while switching to class-based views might not result in noticeable performance gains, the advantages of customization, consistency, documentation, code reuse, and testing are immeasurable. These benefits improve developer collaboration, streamline the development process, and eventually result in a Wagtail admin interface that is more durable and easy to maintain.

**Reply by thibaudcolas** [reply]
@Hadiqamrana I think understanding Wagtail is a requirement for all of the projects we are proposing :) If this isn’t clear enough I’m not sure what to do.

**Comment by nishantn31** [comment]
I think adopting class based views across Wagtail Admin will be very sustainable because if we implement CBV, we will be organizing data of users/clients in a structured (Object-oriented) manner and provide better resource management by code reusability across Wagtail project. I'm looking for the areas where CBV can be implemented.

**Reply by thibaudcolas** [reply]
@nishantn31 perhaps you could share which areas you’ve looked at so far / how you’re looking?
